chromium_src/third_party/blink/renderer/modules/storage/dom_window_storage.cc

```diff
@@ -8,47 +8,160 @@
+    return nullptr;
 
-BraveDOMWindowStorage::BraveDOMWindowStorage() {}
-BraveDOMWindowStorage::~BraveDOMWindowStorage() {}
+  String session_storage_id = session_storage_namespace->namespace_id() +
+                              String("ephemeral-session-storage");
+  String local_storage_id = session_storage_namespace->namespace_id() +
+                            String("ephemeral-local-storage");
+  supplement = MakeGarbageCollected<EphemeralStorageNamespaces>(
+      StorageController::GetInstance(), session_storage_id, local_storage_id);
+
+  ProvideTo(*page, supplement);
+  return supplement;
+}
 
 // static
-StorageArea* BraveDOMWindowStorage::sessionStorage(LocalDOMWindow& window,
-                                              ExceptionState& exception_state) {
-  auto* storage =
-      DOMWindowStorage::From(window).sessionStorage(exception_state);
-  MaybeClearAccessDeniedException(storage, window, &exception_state);
-  return storage;
+const char BraveDOMWindowStorage::kSupplementName[] = "DOMWindowStorage";
+
+// static
+BraveDOMWindowStorage& BraveDOMWindowStorage::From(LocalDOMWindow& window) {
+  BraveDOMWindowStorage* supplement =
+      Supplement<LocalDOMWindow>::From<BraveDOMWindowStorage>(window);
+  if (!supplement) {
+    supplement = MakeGarbageCollected<BraveDOMWindowStorage>(window);
+    ProvideTo(window, supplement);
+  }
+  return *supplement;
 }
 
 // static
-StorageArea* BraveDOMWindowStorage::localStorage(LocalDOMWindow& window,
-                                            ExceptionState& exception_state) {
+StorageArea* BraveDOMWindowStorage::sessionStorage(
+    LocalDOMWindow& window,
+    ExceptionState& exception_state) {
+  return From(window).sessionStorage(exception_state);
+}
+
+// static
+StorageArea* BraveDOMWindowStorage::localStorage(
+    LocalDOMWindow& window,
+    ExceptionState& exception_state) {
+  return From(window).localStorage(exception_state);
+}
+
+BraveDOMWindowStorage::BraveDOMWindowStorage(LocalDOMWindow& window)
+    : Supplement<LocalDOMWindow>(window) {}
+
+StorageArea* BraveDOMWindowStorage::sessionStorage(
+    ExceptionState& exception_state) const {
+  if (ephemeral_session_storage_)
+    return ephemeral_session_storage_;
+
+  LocalDOMWindow* window = GetSupplementable();
   auto* storage =
-      DOMWindowStorage::From(window).localStorage(exception_state);
-  MaybeClearAccessDeniedException(storage, window, &exception_state);
-  return storage;
+      DOMWindowStorage::From(*window).sessionStorage(exception_state);
+
+  if (!window->IsCrossSiteSubframe() || !storage)
+    return storage;
+
+  Page* page = window->GetFrame()->GetDocument()->GetPage();
+  EphemeralStorageNamespaces* namespaces =
+      EphemeralStorageNamespaces::From(page);
+  if (!namespaces)
+    return nullptr;
+
+  Document* document = window->GetFrame()->GetDocument();
+  auto storage_area =
+      namespaces->local_storage()->GetCachedArea(document->GetSecurityOrigin());
+  ephemeral_session_storage_ =
```

Comment:
you are always initializing this with the local storage namespace and you only have one storage area (ephemeral_session_storage_), but session and local storage need to use separate storage areas

Reply:
it really seems like you're overcomplicating things here by instantiating BraveDOMWindowStorage and having a local var. It seems like it would be better to rename `EphemeralStorageNamespaces` -> `EphemeralDOMWindowStorage` and have it own the StorageAreas

Reply:
I'm not sure I understand your first comment, but I can see if it's possible to simplify this.

Reply:
The reason why there are two classes here, which is reflected in the original design in blink is that one class is responsible for passing down the `StorageNamespace` from the tab (`Page`) to all of the `DOMWindow`s. This is why `EphemeralStorageNamespaces` is a `Page` supplement and `BraveDOMWindowStorage` is a `DOMWindow` supplement. I believe the reason for this is, like the previous comment, we cannot simply reach into `LocalFrame` here. I'm not sure it is possible to make the `StorageNamespace` per-DOMWindow without:
 1. Breaking an assumption about OOPI
 2. Continuing to store the namespace id as a string as a page `Page` and then duplicating `StorageNamespace` instances per-`DOMWindow`. This doesn't seem like a better solution. Thoughts?

Reply:
> you are always initializing this with the local storage namespace and you only have one storage area (ephemeral_session_storage_), but session and local storage need to use separate storage areas

I'll fix this. Thanks for pointing this out.